src/systems/logical_audio_sensor_plugin/LogicalAudioSensorPlugin.cc

```diff
@@ -0,0 +1,514 @@
+      source.falloffDistance);
+  source.emissionVolume = volumeLevel;
+  logical_audio::validateVolumeLevel(source.emissionVolume);
+  _ecm.CreateComponent(entity, components::LogicalAudioSource(source));
+
+  // save the source's pose as a component
+  _ecm.CreateComponent(entity,
+      components::Pose(pose));
+
+  // save the source's playing information as a component
+  logical_audio::SourcePlayInfo playInfo;
+  playInfo.playing = playing;
+  playInfo.playDuration = std::chrono::seconds(playDuration);
+  _ecm.CreateComponent(entity,
+      components::LogicalAudioSourcePlayInfo(playInfo));
+
+  // create service callbacks that allow this source to be played/stopped
+  std::function<bool(ignition::msgs::Boolean &)> playSrvCb =
+    [this, entity](ignition::msgs::Boolean &_resp)
+    {
+      std::lock_guard<std::mutex> lock(this->playSourceMutex);
+      this->sourceEntities[entity].first = true;
+      _resp.set_data(true);
+      return true;
+    };
+  std::function<bool(ignition::msgs::Boolean &)> stopSrvCb =
+    [this, entity](ignition::msgs::Boolean &_resp)
+    {
+      std::lock_guard<std::mutex> lock(this->stopSourceMutex);
+      this->sourceEntities[entity].second = true;
+      _resp.set_data(true);
+      return true;
+    };
+
+  // create services for this source
+  const auto full_name = scopedName(entity, _ecm);
+  if (!this->node.Advertise(full_name + "/play", playSrvCb))
+  {
+    ignerr << "Error advertising the play source service for source "
+      << id << " in entity " << _parent << ". " << kSourceSkipMsg;
+    return;
+  }
+  if (!this->node.Advertise(full_name + "/stop", stopSrvCb))
+  {
+    ignerr << "Error advertising the stop source service for source "
+      << id << " in entity " << _parent << ". " << kSourceSkipMsg;
+    return;
+  }
+
+  this->sourceEntities.insert({entity, {false, false}});
+}
+
+//////////////////////////////////////////////////
+void LogicalAudioSensorPluginPrivate::CreateMicrophone(
+    const sdf::ElementPtr &_elem,
+    const Entity &_parent,
+    EntityComponentManager &_ecm,
+    SdfEntityCreator &_sdfEntityCreator,
+    std::unordered_set<unsigned int> &_ids)
+{
+  static const std::string kMicSkipMsg =
+    "Skipping the creation of this microphone.\n";
+
+  if (!_elem->HasElement("id"))
+  {
+    ignerr << "Microphone is missing an id. " << kMicSkipMsg;
+    return;
+  }
+  const auto id = _elem->Get<unsigned int>("id");
+
+  // make sure no other microphones exist with the same ID in the parent
+  if (_ids.find(id) != _ids.end())
+  {
+    ignerr << "The specified microphone ID of " << id << " already exists for "
+      << "another microphone in entity " << _parent << ". " << kMicSkipMsg;
+    return;
+  }
+  _ids.insert(id);
+
+  if (!_elem->HasElement("pose"))
+  {
+    ignerr << "Microphone is missing a pose. " << kMicSkipMsg;
+    return;
+  }
+  const auto pose = _elem->Get<ignition::math::Pose3d>("pose");
+
+  if (!_elem->HasElement("volume_threshold"))
```

Comment:
Can the pose and volume threshold be optional and default to zero if not specified?

Reply:
Pose and volume threshold have been made optional in https://github.com/ignitionrobotics/ign-gazebo/pull/401/commits/314477419d2aa946f384204dc99b17d9fcd963b3. I also updated the tutorial in https://github.com/ignitionrobotics/ign-gazebo/pull/401/commits/4aa354effa929d22188f923e00835bc3b8bd3ce8 so that users are aware these parameters are optional and will default to 0 if not specified.